pgx-pg-sys/src/submodules/setjmp.rs

```diff
@@ -45,17 +45,23 @@ being converted into a transaction-aborting Postgres `ERROR` by PGX.
 
 **/
 #[inline(always)]
+#[track_caller]
 pub(crate) unsafe fn pg_guard_ffi_boundary<T, F: FnOnce() -> T>(f: F) -> T {
     // SAFETY: Caller promises not to call us from anything but the main thread.
     unsafe { pg_guard_ffi_boundary_impl(f) }
 }
 
 #[cfg(not(feature = "postgrestd"))]
 #[inline(always)]
+#[track_caller]
 unsafe fn pg_guard_ffi_boundary_impl<T, F: FnOnce() -> T>(f: F) -> T {
     //! This is the version that uses sigsetjmp and all that, for "normal" Rust/PGX interfaces.
     use crate as pg_sys;
 
+    // The next code is definitely thread-unsafe (it manipulates statics in an
+    // unsynchronized manner), so we may as well check here.
+    super::thread_check::check_active_thread();
```

Comment:
Can we `#[cfg]` this out of `--release` builds?  Or make it an `assert!()` or have it behind a feature flag that's disabled by default or something?

We really don't need additional overhead for every call back into Postgres for something that's most likely not going to happen in the typical pgx extension.

Reply:
Hmm, well, it's still unsound. I think we could have a feature to disable this

That said I could also just make this be a single load from `%fs:0`/`tpidr_el0`, as mentioned in the PR. In which case it should only be a few-cycle load and a (easily predicted) branch in the worst case, likely not much worse than array bounds checks.

I'm more concerned that this might break the `fork` case, which I need to do some testing with.

Reply:
My view is that the ideal situation is that it's compiled out of `--release` builds.

The reasoning here is that basically all pgx extensions are going to be single-threaded, just because Postgres is.  If an extension does threading, it'd likely be "accidentally" from some 3rd-party crate like rayon or crossbeam.

Keeping the check for debug builds would, in a perfect world, catch these problems during development and tests.

If we want a feature like `pedantic-threadness-checks` (or whatever) that can be enabled by the developer for `--release` builds, that'd be fine.  I do see the usefulness of this in a release build for overly-paranoid runtime environments.

Reply:
I'm not sure. It seems pretty easy to accidentally end up calling into PGX from multiple threads given that so much of rust assumes thread-safety by default (outside of unsafe)...

> EDIT: As an aside, what kind of runtime overhead does #[track_caller] add per call?

Adds a secret argument of an `&'static Location<'static>`, which gets loaded from static data and passed in. Concretely, less than passing an extra `&'static str` argument.

Reply:
Well.  Then I'd like to see some kind of "benchmark" to see if this shows up.

And I don't really mean anything scientific here.  Like, does calling this function (https://github.com/tcdi/pgx/blob/d1ffdb2969f414652d48548341e9d9ff9d02e672/pgx-examples/srf/src/lib.rs#L15) through SQL as something like `SELECT srf.generate_series(1, 10000000);` show any measurable difference with and without this check?

Where I'm concerned about overhead is in a case like this where a `#[pg_extern]` function is being called 10s-or-100s-of-millions-of-times by some SQL statement.

Reply:
Sure, a request for a benchmark is totally reasonable.

Reply:
For example, the Timescale folks have mentioned `setjmp` overhead already.  My intuition says this would pop up on their radar too.  ZomboDB too, for that matter.

If it doesn't turn out to be significant, then great!

Reply:
This should be significantly cheaper than the call to sigsetjmp, but I'll take the time to ensure that.

Reply:
I measured this in a few ways, both as standalone benchmarks, benchmarks that include the rest of the pgx sigsetjmp error-handling dance, and in real queries (`SELECT my_generate_series(0, 10000000, 1)`, with `\timing` on). I did this on linux and macOS, which reported similar numbers (although Linux numbers were like 10x slower across the board due to the machine being literally the cheapest laptop I could find 3y ago).

The check in this PR takes around 700ps standalone, and increases the average cost the error handling from ~6ns to ~6.5ns in my benchmarks (there's probably around 200ps of overhead from measurement).

When used in a query, these checks had no reliable overhead that I could detect. With or without the thread safety check, `SELECT my_generate_series(0, 10000000, 1)`, reported times between 950ms and 975ms (well, 4.5-5ms on my linux box). I tested this pretty thoroughly (several DBs/extensions), and am confident that it's not a case of the old code remaining in use.

We could get this down to essentially 0-cost[^1] via the use of inline asm. I have implemented this in the `thomcc/threadsafe-asm` branch (the relevant code is https://github.com/tcdi/pgx/blob/thomcc/threadsafe-asm/pgx-pg-sys/src/submodules/thread_check.rs#L115-L194). It wouldn't be *that* hard for me to maintain (I only implemented versions I understood), but is definitely way worse than the version in this PR, so I don't know that I think this is worth it, since the change is not measurable in queries[^2].

It's definitely your call, though -- I'd be totally fine to pull that into this PR if you'd rather take the optimization.

[^1]: Small enough to be comparable with `black_box(0)` on its own (100ps), and to cause no measurable difference when run as part of the other pg error code.

[^2]: I wouldn't have bothered with this had I realized that, but my initial measurements accidentally compared versions with different compiler flags, so I thought the overhead was much more significant at first.